cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZooKeeperRoller.java

```diff
@@ -53,60 +57,61 @@ public Future<Void> maybeRollingUpdate(Reconciliation reconciliation, Labels sel
                         Promise<Void> promise = Promise.promise();
-                        Future<String> leaderFuture = leaderFinder.findZookeeperLeader(reconciliation, podsToRoll.keySet(), clusterCaSecret, coKeySecret);
+                        Future<String> leaderFuture = leaderFinder.findZookeeperLeader(reconciliation, clusterRollContext.podNames(), clusterCaSecret, coKeySecret);
 
                         leaderFuture.compose(leader -> {
                             LOGGER.debugCr(reconciliation, "Zookeeper leader is " + (ZookeeperLeaderFinder.UNKNOWN_LEADER.equals(leader) ? "unknown" : "pod " + leader));
                             Future<Void> fut = Future.succeededFuture();
 
                             // Then roll each non-leader pod => the leader is rolled last
-                            for (Map.Entry<String, List<String>> podEntry : podsToRoll.entrySet())  {
-                                if (podEntry.getValue() != null && !podEntry.getValue().isEmpty()) {
-                                    if (!podEntry.getKey().equals(leader)) {
-                                        LOGGER.debugCr(reconciliation, "Restarting non-leader pod {}", podEntry.getKey());
-                                        // roll the pod and wait until it is ready
-                                        // this prevents rolling into faulty state (note: this applies just for ZK pods)
-                                        fut = fut.compose(ignore -> restartPod(reconciliation, podEntry.getKey(), podEntry.getValue()));
-                                    } else {
-                                        LOGGER.debugCr(reconciliation, "Deferring restart of leader {}", podEntry.getKey());
+                            for (ZookeeperPodContext podContext : clusterRollContext.getPodContextsWithNonReadyFirst())  {
+                                final String podName = podContext.getPodName();
+                                if (podContext.requiresRestart() && !podName.equals(leader)) {
+                                    LOGGER.debugCr(reconciliation, "Restarting non-leader pod {}", podName);
+                                    // roll the pod and wait until it is ready
+                                    // this prevents rolling into faulty state (note: this applies just for ZK pods)
+                                    fut = fut.compose(ignore -> restartPod(reconciliation, podName, podContext.reasonsToRestart));
+                                } else {
+                                    if (podContext.requiresRestart()) {
+                                        LOGGER.debugCr(reconciliation, "Deferring restart of leader {}", podName);
                                     }
+                                    LOGGER.debugCr(reconciliation, "Pod {} does not need to be restarted", podName);
+                                    LOGGER.debugCr(reconciliation, "Waiting for non-restarted pod {} to become ready", podName);
```

Comment:
Did you checked the log messages in a real pod? Doesn't the `readiness` method cover the logging?

Reply:
This bit of logging was following the KafkaRoller logging, I figured it would be good if the logs were similar. 

Looking at it again the ZookeeperRoller logging is misleading anyway like it logs about `Restarting` while it's looping and constructing the composed future, it's not actually restarting pods.

For example in the test where no pods are restarted you see:

```
Pod my-cluster-zookeeper-0 should be rolled due to [Should restart]
Pod my-cluster-zookeeper-1 does not need to be rolled
Pod my-cluster-zookeeper-2 should be rolled due to [Should restart]
Zookeeper leader is pod my-cluster-zookeeper-2
Pod my-cluster-zookeeper-1 does not need to be restarted
 Waiting for non-restarted pod my-cluster-zookeeper-1 to become ready
Restarting non-leader pod my-cluster-zookeeper-0
Deferring restart of leader my-cluster-zookeeper-2
Pod my-cluster-zookeeper-2 does not need to be restarted
 Waiting for non-restarted pod my-cluster-zookeeper-2 to become ready
 ```

Should we move the logging into the futures?

Reply:
Yeah, but the tests don't call the real methods, right? You use mocks. The `readiness` method (and the methods it calls) IMHO logs as well on a debug level. So I wonder if this really adds something new.

Reply:
Yeah it doesn't add anything, I'll remove that log. 

But what about the ZookeeperRoller logging in general? It's writing out misleading logs like in the above: `Restarting non-leader pod my-cluster-zookeeper-0` is logged while building up the composed future but we never restart due to an earlier failure in the future chain.

Reply:
Hmm, I never noticed that. But I guess you can fix that if you want.

Reply:
Since it's not new and shuffling the logs around into futures will complicate this PR more I'd put it off to another PR. Maybe a quick fix for now would be to change the messages a little to indicate it's a planned action.

Reply:
Actually, we could just delete it since the restart future info logs:
```
        LOGGER.infoCr(reconciliation, "Rolling pod {} due to {}", podName, reasons);
```

Reply:
Well, some of them such as the readiness seems new. But in general it is fine to leave the old issues for another PR of course.

Reply:
Okay pushed up logging changes. Now on the first iteration of the pods you will get one of:

- Pod x needs to be restarted
- Deferring restart of leader x
- Pod x does not need to be restarted

logged for each pod.

Reply:
Sounds good. Thanks.